server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Rights.scala

```diff
@@ -0,0 +1,147 @@
+ * *
+ * http://www.apache.org/licenses/LICENSE-2.0                 *
+ * *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+
+package org.apache.james.jmap.mail
+
+import org.apache.james.core.Username
+import org.apache.james.mailbox.model.MailboxACL
+import org.apache.james.mailbox.model.MailboxACL.{EntryKey, Rfc4314Rights}
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.mutable
+import scala.jdk.CollectionConverters._
+
+object Right {
+  val UNSUPPORTED: Option[Boolean] = None
+
+  val Administer = Right(MailboxACL.Right.Administer)
+  val Expunge = Right(MailboxACL.Right.PerformExpunge)
+  val Insert = Right(MailboxACL.Right.Insert)
+  val Lookup = Right(MailboxACL.Right.Lookup)
+  val Read = Right(MailboxACL.Right.Read)
+  val Seen = Right(MailboxACL.Right.WriteSeenFlag)
+  val DeleteMessages = Right(MailboxACL.Right.DeleteMessages)
+  val Write = Right(MailboxACL.Right.Write)
+
+  private val allRights = Seq(Administer, Expunge, Insert, Lookup, Read, Seen, DeleteMessages, Write)
+
+  def forRight(right: MailboxACL.Right): Option[Right] = allRights.find(_.right.equals(right))
+
+  def forChar(c: Char): Option[Right] = allRights.find(_.asCharacter == c)
+}
+
+sealed case class Right(right: MailboxACL.Right) {
+  def asCharacter: Char = right.asCharacter
+
+  def toMailboxRight: MailboxACL.Right = right
+}
+
+object Rights {
+  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[Rights])
+
+  val EMPTY = new Rights(Map())
+
+  class Builder() {
+    private var rights: Seq[(Username, Right)] = Seq()
+
+    def delegateTo(username: Username, rights: Right): Rights.Builder = {
+      this.rights = this.rights :+ (username, rights)
+      this
+    }
+
+    def delegateTo(username: Username, rights: Seq[Right]): Rights.Builder = {
+      this.rights = this.rights ++ rights.map((username, _))
+      this
+    }
+
+    def combine(builder: Rights.Builder): Rights.Builder = {
+      this.rights = this.rights ++ builder.rights
+      this
+    }
+
+    def build: Rights = new Rights(rights.toSeq
+      .groupMap(_._1)(_._2))
+  }
+
+  def builder: Rights.Builder = new Rights.Builder
+
+  def fromACL(acl: MailboxACL): Rights = acl.getEntries.asScala
+    .filter {
+      case (entryKey, _) => isSupported(entryKey)
+    }
+    .map(toRightsBuilder _)
+    .iterator
+    .reduceOption(_ combine _)
+    .getOrElse(builder)
+    .build
+
+  private def toRightsBuilder(entry: (MailboxACL.EntryKey, MailboxACL.Rfc4314Rights)): Rights.Builder = builder
```

Comment:
I don't understand this remark

Reply:
it doesn't really make sense for a method to take a tuple, usually, you pass it several arguments instead. Here, you did that to avoid having to go from a tuple to several arguments for that method and I don't think it's a good decision. A function signature should not be so dependent from a calling context.